### PR TITLE
Change to use pattern matching

### PR DESCRIPTION
**Pull request by teinarss**

### CODE REVIEW

File: OpenRA.Mods.Common/Activities/Attack.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Attack.cs]: Spaces -> Tabs.

**Review by abcdefg30 — DISMISSED** [review]

File: OpenRA.Game/Input/Hotkey.cs
Comment (pchote) [inline OpenRA.Game/Input/Hotkey.cs:84]: does `Hotkey` vs `Hotkey?` matter here?

File: OpenRA.Game/Primitives/float3.cs
Comment (pchote) [inline OpenRA.Game/Primitives/float3.cs:58]: Likewise here.

File: OpenRA.Game/Widgets/Widget.cs
Comment (pchote) [inline OpenRA.Game/Widgets/Widget.cs]: Can we invert these cases to make the code cleaner?

```csharp
if (LoadWidget(id, parent, args) is T widget)
	return widget;

throw new InvalidOperationException("Widget {0} is not of type {1}".F(id, typeof(T).Name));
```

File: OpenRA.Game/World.cs
Comment (pchote) [inline OpenRA.Game/World.cs]: `var rc = (OrderManager.Connection as EchoConnection)?.Recorder;`?

File: OpenRA.Game/Support/ExceptionHandler.cs
Comment (pchote) [inline OpenRA.Game/Support/ExceptionHandler.cs]: `else if (ex is OutOfMemoryException)`?

File: OpenRA.Game/World.cs
Comment (pchote) [inline OpenRA.Game/World.cs]: `if (!(e is ISpatiallyPartitionable))`?

File: OpenRA.Game/World.cs
Comment (pchote) [inline OpenRA.Game/World.cs]: `if (!(e is ISpatiallyPartitionable))`?

File: OpenRA.Mods.Cnc/Traits/SupportPowers/GrantPrerequisiteChargeDrainPower.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/SupportPowers/GrantPrerequisiteChargeDrainPower.cs]: It would take some intentional hackery gymnastics for anyone to create a situation where `Info` isn't `GrantPrerequisiteChargeDrainPowerInfo`, so IMO we should just drop the check and go with the clearer:
```csharp
if (!Active)
	return null;

var info = (GrantPrerequisiteChargeDrainPowerInfo)Info;
return active ? info.ActiveText : available ? info.AvailableText : null;
```

File: OpenRA.Mods.Cnc/Traits/SupportPowers/GrantPrerequisiteChargeDrainPower.cs
Comment (pchote) [inline OpenRA.Mods.Cnc/Traits/SupportPowers/GrantPrerequisiteChargeDrainPower.cs]: Likewise here.

File: OpenRA.Mods.Common/ActorInitializer.cs
Comment (pchote) [inline OpenRA.Mods.Common/ActorInitializer.cs]: These two `if`s can be squashed together:
```csharp
if (destinationType == typeof(string) && value is ActorInitActorReference reference)
	return reference.InternalName;
```

File: OpenRA.Mods.Common/Traits/World/EditorCursorLayer.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/EditorCursorLayer.cs]: How about `actorCenterOffset = (ios as BuildingInfo)?.CenterOffset(world) ?? WVec.Zero;`?

File: OpenRA.Mods.Common/Traits/World/Locomotor.cs
Comment (pchote) [inline OpenRA.Mods.Common/Traits/World/Locomotor.cs]: This can simplify right down to `return self.OccupiesSpace is Mobile;`.

File: OpenRA.Mods.Common/UtilityCommands/UpdateMapCommand.cs
Comment (pchote) [inline OpenRA.Mods.Common/UtilityCommands/UpdateMapCommand.cs]: This is a bit too ugly for my taste.
Can we at least keep a `var folder = new Folder()` before the if?

File: OpenRA.Platforms.Default/Sdl2PlatformWindow.cs
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2PlatformWindow.cs]: Can we invert these two cases so that we're using `c` inside the `if` instead?

File: OpenRA.Game/Input/Hotkey.cs
Comment (teinarss) [reply OpenRA.Game/Input/Hotkey.cs:84]: Never seen it done like this, usually its done like `return obj is Hotkey o && Equals(o);`

**Review by pchote — DISMISSED** [review]

**Comment by pchote** [comment]
Can we get a second review and merge here before we get more rebase issues?

**Review by reaperrr — APPROVED** [review]

**Comment by reaperrr** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/88d6a7e166c49626b36011ee2a1b9acbde87cd8a...e6ed27ddef788582d0f98bbe37abd2cc20369ac9

**Pull request merged**
